transport/stream.go

```diff
@@ -152,3 +156,29 @@ var _ StreamDialer = (*FuncStreamDialer)(nil)
 func (f FuncStreamDialer) DialStream(ctx context.Context, addr string) (StreamConn, error) {
 	return f(ctx, addr)
 }
+
+// WrapConnFuncStreamDialer is a [StreamDialer] that uses the WrapConn function to wrap the [StreamConn]
+// dialed by the Dialer.
+type WrapConnFuncStreamDialer struct {
```

Comment:
This needs more thought. Please make it private if you need for this PR.

First of all, there's very little need for this the way it is with a FuncStreamDialer.

But also, I need something similar for the config but it's more nuanced. For instance:
- It's better if it's a function rather than public structure, so the function can avoid some nesting, like we do with WrapConn.
- The dialer returned by WrapDialer should have a list of StreamConn wrappers, to avoid nesting.
- The StreamConn wrappers need the dialed address. For example, TLS for SNI, or Websocket for URL host. Otherwise you will end up with the IP address, not the host name. So you must pass that to the wrapper
- We probably need a context as well, depending on how we want the implementation to behave. For instance, do we want the TLS handshake to happen on dial, or first read/write?

Given the complexities, let's not introduce the WrapConnFuncStreamDialer in this PR.

Reply:
Sounds good for now. Though it seems to be a common pattern in the repository (especially when fixing the #195 ).